src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java

```diff
@@ -209,34 +225,114 @@ private void loadConfig(Map<String, Object> config, AmazonCloudWatch client) {
+            if (!yamlAwsTagSelect.containsKey("resource_type_selection") || !yamlAwsTagSelect.containsKey("resource_id_dimension") ||
+              !yamlAwsTagSelect.containsKey("tag_selections")) {
+              throw new IllegalArgumentException("Must provide resource_type_selection, resource_id_dimension and tag_selections");
+            }
+            AWSTagSelect awsTagSelect = new AWSTagSelect();
+            rule.awsTagSelect = awsTagSelect;
+
+            awsTagSelect.resourceTypeSelection = (String)yamlAwsTagSelect.get("resource_type_selection");
+            awsTagSelect.resourceIdDimension = (String)yamlAwsTagSelect.get("resource_id_dimension");
+            awsTagSelect.tagSelections = (Map<String, List<String>>)yamlAwsTagSelect.get("tag_selections");
+          }
         }
 
-        loadConfig(rules, client);
+        loadConfig(rules, cloudWatchClient, taggingClient);
     }
 
-    private void loadConfig(ArrayList<MetricRule> rules, AmazonCloudWatch client) {
+    private void loadConfig(ArrayList<MetricRule> rules, AmazonCloudWatch cloudWatchClient, AWSResourceGroupsTaggingAPI taggingClient) {
         synchronized (activeConfig) {
-            activeConfig.client = client;
+            activeConfig.cloudWatchClient = cloudWatchClient;
+            activeConfig.taggingClient = taggingClient;
             activeConfig.rules = rules;
         }
     }
 
+    private AWSCredentialsProvider getRoleCredentialProvider(Map<String, Object> config) {
+        STSAssumeRoleSessionCredentialsProvider credentialsProvider = new STSAssumeRoleSessionCredentialsProvider.Builder(
+                (String) config.get("role_arn"),
+                "cloudwatch_exporter"
+              ).build();
+        return credentialsProvider;
+    }
+    
+    private Region getRegion(Map<String, Object> config) {
+        Region region = RegionUtils.getRegion((String) config.get("region"));
+        if (region == null) {
+          region = Regions.getCurrentRegion();
+          if (region == null) {
+            throw new IllegalArgumentException("No region provided and EC2 metadata failed");
+          }
+        }
+        return region;
+    }
+    
     public String getMonitoringEndpoint(Region region) {
-      return "https://" + region.getServiceEndpoint("monitoring");
+        return "https://" + region.getServiceEndpoint("monitoring");
     }
 
-    private List<List<Dimension>> getDimensions(MetricRule rule, AmazonCloudWatch client) {
+    private List<ResourceTagMapping> getResourceTagMappings(MetricRule rule, AWSResourceGroupsTaggingAPI taggingClient) {
+	List<ResourceTagMapping> resourceTagMappings = new ArrayList<ResourceTagMapping>();
+	if(rule.awsTagSelect == null) {
+            return resourceTagMappings;
+        }
+	
+        List<TagFilter> tagFilters = new ArrayList<TagFilter>();
+        for (Map.Entry<String, List<String>> entry : rule.awsTagSelect.tagSelections.entrySet()) {
```

Comment:
Is the entrySet thread safe ? I think this comes from the config so if it's parsed only at launch time then I'm good with it. Otherwise we should use a Hashtable or something else

Reply:
The map is never modified , it is `created` at config parsing, no I see no problem.